roles/backup/templates/backup-monitoring-alerts.yml.j2

```
@@ -0,0 +1,40 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    monitoring-key: middleware
+  name: backup-monitoring-alerts
+spec:
+  groups:
+    - name: general.rules
+      rules:
+      - alert: JobRunningTimeExceeded
+        annotations:
+          message: Job {{ '{{' }} $labels.namespace {{ '}}' }}/{{ '{{' }} $labels.job {{ '}}' }} has been running for longer than 5 minutes
+        expr: |
+          time() - (max(kube_job_status_active * ON(job) GROUP_RIGHT() kube_job_labels{label_monitoring_key="middleware"}) BY (job) * ON(job) GROUP_RIGHT() max(kube_job_status_start_time * ON(job) GROUP_RIGHT() kube_job_labels{label_monitoring_key="middleware"}) BY (job, namespace, label_cronjob_name) > 0) > 300
+        labels:
+          severity: critical
+      - alert: CronJobSuspended
+        annotations:
+          message: CronJob {{ '{{' }} $labels.namespace {{ '}}' }}/{{ '{{' }} $labels.cronjob {{ '}}' }} is suspended
+        expr: |
+          kube_cronjob_labels{ label_monitoring_key="middleware" } * ON (cronjob) GROUP_RIGHT() kube_cronjob_spec_suspend > 0
+        for: 60s
+        labels:
+          severity: critical
+      - alert: CronJobNotRunInThreshold
+        annotations:
+          message: CronJob {{ '{{' }} $labels.namespace {{ '}}' }}/{{ '{{' }} $labels.label_cronjob_name {{ '}}' }} has not started a Job in
+        expr: |
+          (time() - (max( kube_job_status_start_time * ON(job) GROUP_RIGHT() kube_job_labels{label_monitoring_key="middleware"} ) BY (job, label_cronjob_name) == ON(label_cronjob_name) GROUP_LEFT() max( kube_job_status_start_time * ON(job) GROUP_RIGHT() kube_job_labels{label_monitoring_key="middleware"} ) BY (label_cronjob_name))) > 60*60*25
+      {% for cronjob in expected_cronjobs %}
```

@aidenkeating I'm not sure about this approach for building up the cronjob alerts.
If you take the following scenario:

* install runs, but the fuse online cronjob wasn't installed for some reason (error swallowed, or the cronjob was accidentally removed from the installer)
* alerts are added for all cronjobs found, which doesn't include fuse online.
* we don't get any alerts about the missing cronjob

I'd prefer if we were explicit about the jobs that should exist.
This also makes it easier & more noticeable for upgrades if a new cronjob was added i.e. this file will be updated and show up in a diff.

---

@david-martin Sure, makes sense. I was coming from the other point-of-view, that we could end up with an unmaintained list of CronJob names and when new CronJobs were added (especially in operators) that they wouldn't be added to a list in the installer.

I'll update this to a static list.

---

> when new CronJobs were added (especially in operators) that they wouldn't be added to a list

This is a good point. I'd like to see how that plays out with an operator. In those cases do we trust the operator takes care of backups given the right aws credentials, or maybe the operator also includes a cronjob alert check in a PrometheusRule in that service's namespace.

@pb82 @philbrookes  Is this something you've thought about for the keycloak operator?

---

This would currently not work for the keycloak backup job because it does not label it's CronJob with `monitoring-key: middleware`: https://github.com/integr8ly/keycloak-operator/blob/master/pkg/keycloak/phaseHandler.go#L223

ping @philbrookes

---

@pb82 I am working on this change now (to add labels to the CronJob and the Jobs it creates), as it's needed for the other general checks anyway.

---

> @aidenkeating I'm not sure about this approach for building up the cronjob alerts.
> If you take the following scenario:
> 
> * install runs, but the fuse online cronjob wasn't installed for some reason (error swallowed, or the cronjob was accidentally removed from the installer)
> * alerts are added for all cronjobs found, which doesn't include fuse online.
> * we don't get any alerts about the missing cronjob
> 
> I'd prefer if we were explicit about the jobs that should exist.
> This also makes it easier & more noticeable for upgrades if a new cronjob was added i.e. this file will be updated and show up in a diff.

If you are scraping the known namespaces and creating cronjobs based on this it stands to reason that you will have an alert to ensure all alerts have been created based off the namespaces themselves, rather than the cronjobs in existence. 

Could it be done, easily, that the install can specify the number of cronjobs expected if backups are enabled and exposing that in a label or similar? That way, for a given namespace with a given label X alerts and cronjobs should exist. An alert can be created to then ensure this number matches the info scraped from the namespaces.

---

@pat-cremin From what I understand, the question is whether we can have a mapping between namespace names and the amount of CronJobs we expect to be in those namespaces (via a label)? For example:

The only concern I have with that is that when the alert kicks off we won't then know what CronJob isn't there. A person seeing this alert would know that a CronJob is missing in a namespace. They'd have to check the CronJobs in that namespace and have some documentation about what CronJobs should exist there (I don't know if we have that).

IMO, if we're going to have a static list we may as well specify the names of each CronJob expected and not the count so that when an alert fires a person knows that the MySQL CronJob for SSO is missing without doing any investigation and can restore it using some SOP or something.

As for whether it gets stored as a label or just a var in the installer, I don't have any preference.

Could be missing something in your comment though? Any thoughts?

---

@aidenkeating Your suggestion works just as well. The label or var could allow the inherent list through standardisation of the cron names. eg. 3scale mysql with backup enabled would expect a "3scale-mysql-backup-cron", with label on the component to state it will need a backup, the logic to  create the correct number of crons and know the exact names of them follows organically.

---

@aidenkeating given the new approach to cronjob alerts, I think it's important to document how this works.
I'd like to see at least a followup issue for documenting what's involved in adding a new CronJob for some component, including alerts for that cronjob.

---

@david-martin I have created a follow-up ticket, INTLY-1330, to address this documentation and assigned it to myself.